src/ServiceControl/CompositeViews/Messages/RoutedApi.cs

```diff
@@ -0,0 +1,109 @@
+namespace ServiceControl.CompositeViews.Messages
+{
+    using System;
+    using System.Linq;
+    using System.Net.Http;
+    using System.Threading.Tasks;
+    using Nancy;
+    using NServiceBus.Logging;
+    using Raven.Client;
+    using ServiceBus.Management.Infrastructure.Nancy.Modules;
+    using ServiceBus.Management.Infrastructure.Settings;
+    using ServiceControl.Infrastructure.Settings;
+    using HttpStatusCode = System.Net.HttpStatusCode;
+
+    public abstract class RoutedApi<TIn> : IApi
+    {
+        static ILog logger = LogManager.GetLogger(typeof(RoutedApi<TIn>));
+
+        public IDocumentStore Store { get; set; }
+        public Settings Settings { get; set; }
+        public Func<HttpClient> HttpClientFactory { get; set; }
+
+        public async Task<Response> Execute(BaseModule module, TIn input)
+        {
+            var currentRequest = module.Request;
+            Response response;
+
+            var instanceId = GetInstance(currentRequest, input);
+
+            var localInstanceId = InstanceIdGenerator.FromApiUrl(Settings.ApiUrl);
+
+            if (!string.IsNullOrWhiteSpace(instanceId))
+            {
+                if (instanceId == localInstanceId)
+                {
+                    response = await LocalQuery(currentRequest, input, localInstanceId);
+                }
+                else
+                {
+                    response = await RemoteCall(currentRequest, instanceId);
+                }
+            }
+            else
+            {
+                response = await LocalQuery(currentRequest, input, localInstanceId);
+            }
+
+            return response;
+        }
+
+        protected virtual string GetInstance(Request currentRequest, TIn input)
+        {
+            return (string)currentRequest.Query.instance_id;
+        }
+
+        protected abstract Task<Response> LocalQuery(Request request, TIn input, string instanceId);
+
+        private async Task<Response> RemoteCall(Request currentRequest, string instanceId)
+        {
+            var remoteUri = InstanceIdGenerator.ToApiUrl(instanceId);
+
+            var instanceUri = new Uri($"{remoteUri}{currentRequest.Path}{currentRequest.Url.Query}");
+            var httpClient = HttpClientFactory();
+            try
+            {
+                var method = new HttpMethod(currentRequest.Method);
+                var requestMessage = new HttpRequestMessage(method, instanceUri);
+                foreach (var currentRequestHeader in currentRequest.Headers)
+                {
+                    requestMessage.Headers.Add(currentRequestHeader.Key, currentRequestHeader.Value);
+                }
+
+                var rawResponse = await httpClient.SendAsync(requestMessage).ConfigureAwait(false);
+
+                // TODO: is it correct to merge headers??
```

Comment:
Looks right. `response.Headers` and `response.Content.Headers` should have different headers/keys.